src/encoded/audit/experiment.py

```diff
@@ -57,162 +57,177 @@ def audit_experiment_chipseq_control_read_depth(value, system, files_structure):
         target_name = value['target']['name']
         target_investigated_as = value['target']['investigated_as']
         if target_name not in ['Control-human', 'Control-mouse']:
-            for alignment_file in files_structure.get('alignments').values():
-                # initially was for file award
-                if not alignment_file.get('award') or \
-                    alignment_file.get('award')['rfa'] not in [
-                        'ENCODE3',
-                        'ENCODE4',
-                        'ENCODE2-Mouse',
-                        'ENCODE2',
-                        'ENCODE',
-                        'Roadmap']:
-                    continue
-                if alignment_file.get('lab') not in ['/labs/encode-processing-pipeline/']:
-                    continue
-                derived_from_files = list(
-                    get_derived_from_files_set([alignment_file], files_structure, 'fastq', True))
-                if not derived_from_files:
-                    continue
-                control_bam = get_control_bam(
-                    alignment_file,
-                    'ChIP-seq read mapping',
-                    derived_from_files,
-                    files_structure)
-                if control_bam is not False:
-                    control_depth = get_chip_seq_bam_read_depth(control_bam)
-                    control_target = get_target_name(derived_from_files)
-                    if control_depth is not False and control_target is not False:
-                        yield from check_control_read_depth_standards(
-                            control_bam,
-                            control_depth,
-                            control_target,
-                            True,
-                            target_name,
-                            target_investigated_as)
-    return
+            controls = value.get('possible_controls')
+            if controls:
+                controls_files_structures = {}
+                control_objects = {}
+                for control_experiment in controls:
+                    control_objects[control_experiment.get('@id')] = control_experiment
+                    controls_files_structures[control_experiment.get('@id')] = create_files_mapping(
+                        control_experiment.get('original_files'),
+                        files_structure.get('excluded_types'))
+                for peaks_file in files_structure.get('peaks_files').values():
```

Comment:
All of the peaks_file conditionals under
```
for peaks_file in files_structure.get('peaks_files').values():
```
are filters for files_structure.get('peaks_files').values().  Lines 69 - 83 can be put in a generator expression much like 'derived_from_external_bams' below.  Although, 'derived_from_external_bams' should be a generator expression and not a list comprehension.  In general, if you have a list or set that you need to consume once, then you can make it a generator.

```
peaks_files_gen = (
    peaks_file
    for peak_file in files_structure.get('peaks_files').values()
    ...condition
    ...condition
)
for peaks_file in peaks_files:
    derived_from_files = etc...
    derived_from_external_bams = (
        derived_from
        derived_from in derived_from_files
        ...condition
        ...condition
    )
```

Reply:
require re-demo